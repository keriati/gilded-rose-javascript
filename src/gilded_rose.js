import {
    lensProp,
    over,
    subtract,
    inc,
    propEq,
    dec,
    clamp,
    cond,
    always,
    compose,
    allPass,
    ifElse,
    identity,
} from 'ramda'

export class Item {
    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}


export const TICKET_NAME = 'Backstage passes to a TAFKAL80ETC concert'
export const CHEESE_NAME = 'Aged Brie'
export const SULFURAS = 'Sulfuras, Hand of Ragnaros'
export const CONJURED = 'Conjured Mana Cake'

export const QUALITY = 'quality'
export const SELL_IN = 'sellIn'
export const NAME = 'name'

// helpers, utils
export const nameEqual = propEq(NAME)
export const isCheese = nameEqual(CHEESE_NAME)
export const isTicket = nameEqual(TICKET_NAME)
export const isSulfuras = nameEqual(SULFURAS)
export const isConjured = nameEqual(CONJURED)
export const isCommonItem = (item) => !allPass[isConjured, isSulfuras, isCheese, isTicket]

export const qualityLens = lensProp(QUALITY)
export const overQuality = over(qualityLens)

export const qualityBoundaries = clamp(0, 50)
export const increaseQuality = overQuality(inc)
export const decreaseQuality = overQuality(dec)
export const decreaseQualityBy2 = over(qualityLens, subtract(2))
export const boundQuality = overQuality(qualityBoundaries)

export const sellInLens = lensProp(SELL_IN)
export const decreaseSellIn = over(sellInLens, dec)

export const increaseQualityAfterSellDate = (item) => {
    return item.sellIn < 0 ? increaseQuality(item) : item
}
export const decreaseQualityAfterSellDate = (item) => {
    return item.sellIn < 0 ? decreaseQuality(item) : item
}
export const increaseQualityFor10DaysBeforeSellDate = (item) => {
    return item.sellIn < 11 ? increaseQuality(item) : item
}
export const increaseQualityFor6DaysBeforeSellDate = (item) => {
    return item.sellIn < 6 ? increaseQuality(item) : item
}
export const dropQualityToZeroAfterSellDate = (item) => {
    return item.sellIn < 0 ? overQuality(always(0), item) : item
}


// item updates
export const updateCommonItem = compose(
    decreaseQualityAfterSellDate,
    decreaseQuality,
)
export const updateTicket = compose(
    dropQualityToZeroAfterSellDate,
    increaseQualityFor6DaysBeforeSellDate,
    increaseQualityFor10DaysBeforeSellDate,
    increaseQuality
)
export const updateCheese = compose(
    increaseQualityAfterSellDate,
    increaseQuality,
)
export const updateConjuredItem = item => {
   return isConjured(item) ? decreaseQualityBy2(item) : item
}

// general
export const updateQuality = cond([
    [isConjured, updateConjuredItem],
    [isCheese, updateCheese],
    [isTicket, updateTicket],
    [isCommonItem, updateCommonItem],
])

export const update = compose(
    decreaseSellIn,
    boundQuality,
    updateQuality
)

export class Shop {
    constructor(items = []) {
        this.items = items;
    }

    updateQuality() {
        for (let i = 0; i < this.items.length; i++) {
            const item = this.items[i]
            this.items[i] = isSulfuras(item) ? item : update(item)
        }

        return this.items;
    }
}
