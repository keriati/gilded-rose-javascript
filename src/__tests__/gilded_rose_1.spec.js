import {Shop, Item} from '../gilded_rose';

describe("Guilded rose: ", function () {

    const items = [];

    items.push(new Item('+5 Dexterity Vest', 10, 20));
    items.push(new Item('Aged Brie', 2, 0));
    items.push(new Item('Elixir of the Mongoose', 5, 7));
    items.push(new Item('Sulfuras, Hand of Ragnaros', 0, 80));
    items.push(new Item('Sulfuras, Hand of Ragnaros', -1, 80));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 5, 50));


    for (let i = 0; i <= 5; i++) {
        it("match snapshot after day " + i, function () {
            const gildedRose = new Shop(items);
            const updatedItems = gildedRose.updateQuality();
            expect(updatedItems).toMatchSnapshot();
        })

    }

});
