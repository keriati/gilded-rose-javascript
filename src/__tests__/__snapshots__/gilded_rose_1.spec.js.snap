// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guilded rose:  match snapshot after day 0 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 19,
    "sellIn": 9,
  },
  Item {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
]
`;

exports[`Guilded rose:  match snapshot after day 1 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  Item {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
]
`;

exports[`Guilded rose:  match snapshot after day 2 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 17,
    "sellIn": 7,
  },
  Item {
    "name": "Aged Brie",
    "quality": 4,
    "sellIn": -1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 4,
    "sellIn": 2,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 23,
    "sellIn": 12,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 7,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 2,
  },
]
`;

exports[`Guilded rose:  match snapshot after day 3 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 16,
    "sellIn": 6,
  },
  Item {
    "name": "Aged Brie",
    "quality": 6,
    "sellIn": -2,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 3,
    "sellIn": 1,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 24,
    "sellIn": 11,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 6,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 1,
  },
]
`;

exports[`Guilded rose:  match snapshot after day 4 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 15,
    "sellIn": 5,
  },
  Item {
    "name": "Aged Brie",
    "quality": 8,
    "sellIn": -3,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 25,
    "sellIn": 10,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 5,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 0,
  },
]
`;

exports[`Guilded rose:  match snapshot after day 5 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 14,
    "sellIn": 4,
  },
  Item {
    "name": "Aged Brie",
    "quality": 10,
    "sellIn": -4,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -1,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 27,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
]
`;
